visual-dotnet/SauceLabs.Visual/BuildFactory.cs

```diff
@@ -0,0 +1,172 @@
+            {
+                throw new VisualClientException($@"build {buildId} was not found");
+            }
+        }
+
+        /// <summary>
+        /// <c>FindBuildById</c> returns the build identified by <c>buildId</c>
+        /// </summary>
+        /// <param name="api">a <c>VisualApi</c> object</param>
+        /// <param name="customId">the <c>customId</c> to look for</param>
+        /// <returns>the matching build</returns>
+        /// <exception cref="VisualClientException">when build is not existing or has an invalid state</exception>
+        private static async Task<VisualBuild?> FindBuildByCustomId(VisualApi api, string customId)
+        {
+            try
+            {
+                var build = (await api.BuildByCustomId(customId)).EnsureValidResponse().Result;
+                return new VisualBuild(build.Id, build.Url, build.Mode);
+            }
+            catch (VisualClientException)
+            {
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// <c>CloseBuilds</c> closes all build that are still open.
+        /// </summary>
+        internal static async Task CloseBuilds()
+        {
+            var builds = Builds.Values.ToArray();
+            foreach (var build in builds)
+            {
+                if (build.Close != null)
+                {
+                    await build.Close();
+                }
+            }
+        }
+
+        /// <summary>
+        /// <c>CreateBuild</c> creates a new Visual build.
+        /// </summary>
+        /// <param name="client">the client used for the build creation</param>
+        /// <param name="options">the options for the build creation</param>
+        /// <returns>a <c>VisualBuild</c> instance</returns>
+        private static async Task<VisualBuild> Create(VisualApi api, CreateBuildOptions options)
+        {
+            var build = await GetEffectiveBuild(api, EnvVars.BuildId, EnvVars.CustomId);
+            if (build != null)
+            {
+                if (!build.IsRunning())
+                {
+                    throw new VisualClientException($"build {build.Id} is not RUNNING");
+                }
+
+                build.IsExternal = true;
+                build.Close = async () => Disregard(build);
+                return build;
+            }
+
+            options.CustomId ??= EnvVars.CustomId;
+            var result = (await api.CreateBuild(new CreateBuildIn
+            {
+                Name = options.Name,
+                Project = options.Project,
+                Branch = options.Branch,
+                CustomId = options.CustomId,
+                DefaultBranch = options.DefaultBranch,
+            })).EnsureValidResponse();
+
+            build = new VisualBuild(result.Result.Id, result.Result.Url, result.Result.Mode)
+            {
+                IsExternal = false
+            };
+
+            var copiedApi = api.Clone();
+            build.Close = async () =>
```

Comment:
maybe it's a microoptimization but maybe from the code perspective it would look better. How about not only caching the build but also an HTTP client?
then we don't need to even make a copy of the client. Every time you make a copy you also reserve a port and other resources. Maybe it's worth to also cache it?

Reply:
Correct me if I'm wrong, but until a request has been made, it should not reserve anything network related. A bit of memory yes, but not sure it uses a port at this stage.